feat: Add new privacy policy alert toast

jonybur opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Changes:
Adds a new privacy policy banner that will show up after May 6, 2024. [EDIT HB - date set for June 4, 2023]
If the user closes it manually, the privacy policy banner will not show up again.
If the user does not close the toast manually, the privacy policy banner will cease to show up after a day of being shown for the first time.
Changes the Survey banner so that we can display multiple banners at the same time without overlapping each other.

Considerations:
- Make sure this doesn't show up during onboarding
- This toast will show up even after years of the privacy policy being changed. We should set up a deadline for that or eventually revert this PR.
- Check if copy is correct.

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to the home page
2. Check if there's a new privacy banner there (date should be edited in the code so that the privacy policy shows up earlier than May 6, 2024)
3. When clicking the link, it should take you to the Privacy Policy website
4. When closing it, it shouldn't show up again
5. If not closing it manually, the toast should not show up again after 1 day of being shown for the first time. This data is saved in localStorage so editing the timestamp there should be enough to test this feature out.

## **Screenshots/Recordings**

https://github.com/MetaMask/metamask-mobile/assets/11148144/ae6e74fc-0a7f-47ce-8744-4a292875f7b3


### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Commit:
Add Toast, remove autoclose

Commit:
Add Toast for new privacy policy

Commit:
Move functionality from App to Wallet

Commit:
Do not show new privacy policy toast when onboarding

Commit:
Add close button

Commit:
Move new privacy date to year 2100

hesterbruikman commented:
@jonybur Same as for Extension, could you remove the Accept button here please? It's not required from a legal point of view and it might create the impression that it is

Commit:
Update date

Commit:
Merge branch 'main' into new-privacy-policy-toast

Commit:
Fix copy

Commit:
Merge branch 'main' into new-privacy-policy-toast

georgewrmarshall reviewed (commented):
Code LGTM! Would be great to upload the most recent recording that removes the approve button and it looks like you may have some linting errors and need to sign the CLA @jonybur 🙏

EtherWizard33 commented on app/components/Nav/App/index.js:
```diff
@@ -49,6 +49,7 @@ import { useTheme } from '../../../util/theme';
 import Device from '../../../util/device';
 import SDKConnect from '../../../core/SDKConnect/SDKConnect';
 import { colors as importedColors } from '../../../styles/common';
```
just a nit: any reason the new line at line 52 is added?

jonybur commented:
I have read the CLA Document and I hereby sign the CLA

Commit:
Fix lint

Commit:
Fix root state

EtherWizard33 commented on app/reducers/legalNotices/index.ts:
```diff
@@ -0,0 +1,65 @@
+      (isRecent && !newPrivacyPolicyToastClickedOrClosed))
+  );
+};
+
```
nit: maybe the 'any' could be typed.

It seems in a few reducers they do it inline like  action: { type: string, payload: number}

Or I have also seen an example in the accounts reducer extending the redux Action type:

export interface iAccountActions extends Action {
  reloadAccounts: boolean;
}

EtherWizard33 reviewed (commented):
looks like there is a snapshot test failing, this should create an updated snapshot:

npx jest -u ./app/components/Views/Wallet/index.test.tsx;

Commit:
Add type

Commit:
Merge branch 'main' into new-privacy-policy-toast

Commit:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into new-privacy-policy-toast

Commit:
Merge branch 'new-privacy-policy-toast' of github.com:MetaMask/metamask-mobile into new-privacy-policy-toast

Commit:
Update snapshot

Commit:
Merge branch 'main' into new-privacy-policy-toast

Commit:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into new-privacy-policy-toast

Commit:
Merge branch 'new-privacy-policy-toast' of github.com:MetaMask/metamask-mobile into new-privacy-policy-toast

Commit:
Update snap

EtherWizard33 commented on app/reducers/legalNotices/index.ts:
```diff
@@ -0,0 +1,78 @@
+  );
+};
+
+export interface iAccountActions extends Action {
```
NIT: this is just a nit so feel free to use it or not, I noticed it's not used very much currently in the app anyway.
But if it's used it should probably be named LegalNoticesActions, here is an example, the action typing stuff is between line 18 and 28 of this example, the rest I provide just for a bit of context around the example:

https://chat.consensys.io/chat/OnL8kM75dDqArR335YFxW

EtherWizard33 committed:
test: update snapshot

georgewrmarshall reviewed (dismissed):
Component library changes LGTM! Approving to unblock. Please still get an approval from @EtherWizard33 🙏

Commit:
Update name of action interface

Commit:
Merge branch 'new-privacy-policy-toast' of github.com:MetaMask/metamask-mobile into new-privacy-policy-toast

Commit:
Merge branch 'main' into new-privacy-policy-toast

EtherWizard33 commented:
> Component library changes LGTM! Approving to unblock. Please still get an approval from @EtherWizard33 🙏

Also needed approval from mobile platform code owners first, EtherWizard33 being a newcomer probably shouldn't be really stamping stuff yet.

EtherWizard33 commented on app/components/Views/Wallet/index.tsx:
```diff
@@ -148,6 +176,44 @@ const Wallet = ({ navigation }: any) => {
+    });
+  }, [
+    storePrivacyPolicyShownDate,
+    shouldShowNewPrivacyToast,
```
is the toast supposed to be rendered only once on first rendering, and if it would render only once, would it affect how it's expected to work?

Without dependencies i logged one showToast, and with I logged 3-4 as can be see in attached images 


| With deps       | Without deps        |
|--------------|--------------|
| <img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/MetaMask/metamask-mobile/assets/165834542/46b12768-8bba-4a2c-8a64-a0658d0951c8"> |<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/MetaMask/metamask-mobile/assets/165834542/8c2483e5-7c3b-409b-9dbd-12a106ca5a78"> |

Commit:
feat: Update onboarding copy (#9256)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Updates the onboarding text.

This PR is against the new-privacy-policy-toast since they need to go
together

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2220

## **Manual testing steps**

1. Start onboarding a new user
2. Check that the notice has the new copy

## **Screenshots/Recordings**

<img width="385" alt="Screenshot 2024-04-15 at 14 25 19"
src="https://github.com/MetaMask/metamask-mobile/assets/11148144/dc270589-1b93-4627-bac1-87563afd1086">


### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

EtherWizard33 commented on locales/languages/en.json:
```diff
@@ -500,24 +500,38 @@
-    "fine_print_2b": "before proceeding.\nFor more information on how MetaMask and Infura interact from a data collection perspective, see our update",
-    "fine_print_2c": ". For more information on our privacy practices in general, see our Privacy Policy",
-    "here": "here",
+    "fine_print_1": "We’ll let you know if we decide to use this data for other purposes. You can review our",
```
Are the translations displaying for you, currently I see this:


| en-US       | en-CA        |
|--------------|--------------|
| <img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/MetaMask/metamask-mobile/assets/165834542/3ec3a835-3ff9-4e85-82ab-21366127fe98"> |<img width="350" alt="Screenshot 2024-04-18 at 3 56 43 PM" src="https://github.com/MetaMask/metamask-mobile/assets/165834542/5158bbcc-e06c-4808-a03f-bb25d1a927e5"> |

jonybur replied:
Fixed!

Commit:
Fix legacy strings

Commit:
Fix legacy text

tommasini commented on app/components/Views/Wallet/index.tsx:
```diff
@@ -438,4 +504,15 @@ const Wallet = ({ navigation }: any) => {
 };
 
-export default Wallet;
+const mapStateToProps = (state: any) => ({
```
Instead of using this pattern of mapStateToProps and mapDispatchToProps, since we have a functional component what do you think about using the useSelector (That it's already used ) and the useDispatch hooks instead?

jonybur replied:
Hey @tommasini , this will involve a re-do of large parts of the PR. I would rather not do this, but we can absolutely apply those refactors after merging this PR, given that this must go in ASAP.

Commit:
Fix up feature flag

Commit:
Fix snap

Commit:
Remove duplicate code

Commit:
Fix title

EtherWizard33 reviewed (dismissed):
ran it on android and ios and it showed and was hidden as described in the pr description

approving assuming passed CI checks and considering there is an ongoing review from the platform team as well

Commit:
Merge branch 'main' into new-privacy-policy-toast

brianacnguyen commented on app/component-library/components/Toast/Toast.types.ts:
```diff
@@ -29,12 +29,19 @@ export interface ToastLinkButtonOptions {
+  disableTimeout: boolean;
   labelOptions: ToastLabelOptions;
   linkButtonOptions?: ToastLinkButtonOptions;
+  closeButtonOptions?: ToastCloseButtonOptions;
```
can you update this to use ButtonProps instead to avoid more unnecessary duplicates

brianacnguyen commented on app/component-library/components/Toast/Toast.types.ts:
```diff
@@ -29,12 +29,19 @@ export interface ToastLinkButtonOptions {
  * Common toast option shared between all other options.
  */
 interface BaseToastVariants {
+  disableTimeout: boolean;
```
can you update this with is- or has- 
https://github.com/MetaMask/contributor-docs/blob/main/docs/javascript.md

brianacnguyen reviewed (changes requested):
Requesting a few changes with the DS changes to follow best practices

Commit:
Fix types

Commit:
Merge branch 'new-privacy-policy-toast' of github.com:MetaMask/metamask-mobile into new-privacy-policy-toast

tommasini reviewed (approved):
Approving this so it can go in, in 7.23, but I would say that this PR should change:

On wallet/index.tsx
`connect`, `mapStateToProps` and `mapDispatchToProps` for the hooks (useSelector and useDispatch) of redux since it's a functional component and it was already following that pattern

It would be a nice to have the slices of redux tool kit follwing the slice pattern already implemented instead of creating a new action and reducer


BREAKING:

This PR absolutely needs the translations to be ready to cherry pick into 7.23.0
